Sources/PackageCollectionGenerator/PackageMetadataProviders/GitLabPackageManagerMetadataProvider.swift

```diff
@@ -0,0 +1,170 @@
+                result = .failure(Errors.notFound(metadataURL))
+            case (200, _, _):
+                guard let metadata = try? response.decodeBody(GetRepositoryResponse.self, using: self.decoder) else {
+                    callback(.failure(Errors.invalidResponse(metadataURL, "Invalid body")))
+                    return
+                }
+
+                let license = metadata.license
+                let packageLicense: PackageCollectionModel.V1.License?
+                if let license = license {
+                    packageLicense = .init(name: license.name, url: license.sourceURL)
+                } else if let licenseURL = metadata.license_url,
+                    let licenseURL = URL(string: licenseURL) {
+                    packageLicense = .init(name: nil, url: licenseURL)
+                } else {
+                    packageLicense = nil
+                }
+
+                let model = PackageBasicMetadata(
+                    summary: metadata.description,
+                    keywords: metadata.topics,
+                    readmeURL: metadata.readme_url.flatMap { URL(string: $0) },
+                    license: packageLicense
+                )
+
+                callback(.success(model))
+            default:
+                callback(.failure(Errors.invalidResponse(metadataURL, "Invalid status code: \(response.statusCode)")))
+            }
+        }
+    }
+
+    internal func apiURL(_ url: String) -> Foundation.URL? {
+        return URL(string: url.spm_dropGitSuffix())
+    }
+
+    private func makeRequestOptions(validResponseCodes: [Int]) -> HTTPClientRequest.Options {
+        var options = HTTPClientRequest.Options()
+        options.addUserAgent = true
+        options.validResponseCodes = validResponseCodes
+        options.authorizationProvider = { url in
+            url.host.flatMap { host in
+                return self.authTokens[.gitlab(host)].flatMap { token in "Bearer \(token)" }
+            }
+        }
+        return options
+    }
+
+    private static func makeDefaultHTTPClient() -> HTTPClient {
+        var client = HTTPClient()
+        client.configuration.requestTimeout = .seconds(2)
+        client.configuration.retryStrategy = .exponentialBackoff(maxAttempts: 3, baseDelay: .milliseconds(50))
+        client.configuration.circuitBreakerStrategy = .hostErrors(maxErrors: 50, age: .seconds(30))
+        return client
+    }
+
+    enum Errors: Error, Equatable {
+        case invalidGitURL(URL)
+        case invalidResponse(URL, String)
+        case permissionDenied(URL)
+        case invalidAuthToken(URL)
+        case apiLimitsExceeded(URL, Int, Int)
+        case notFound(URL)
+    }
+}
+
+extension GitLabPackageMetadataProvider {
+    fileprivate struct GetRepositoryResponse: Codable {
+        let name: String
+        let fullName: String
+        let description: String?
+        let topics: [String]?
+        let license_url: String?
+        let readme_url: String?
```

Comment:
Do license and README URLs serve raw contents of the files or they are rendered HTML pages?

For example, the README for this GitHub repo can be viewed at:
1. https://github.com/apple/swift-package-collection-generator/blob/main/README.md
2. https://raw.githubusercontent.com/apple/swift-package-collection-generator/main/README.md 

We want URL no.2 in the package collection.

GitHub has separate APIs for fetching the "download URL" for license and README. I don't know if GitLab has the same? This is important for README especially.

Reply:
It is the URL to the HTML rendered README. There is no raw-link that comes in the data provided from the GitLab-API, but I can build the path by my own (https://docs.gitlab.com/ee/api/repository_files.html#get-raw-file-from-repository). I will do that in another commit.